### PR TITLE
Allow BashOperator to take command as list

### PR DESCRIPTION
**Pull request by dimberman**
---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Review by ashb — COMMENTED** [review]
You should rearange the code so there's only a single `Popen` call - i.e. check then modify bash_command

File: airflow/operators/bash.py [hunk 1]
```diff
@@ -19,16 +19,22 @@
...
+from typing import Dict, List, Optional, Union
 
 from airflow.exceptions import AirflowException
 from airflow.models import BaseOperator
 from airflow.utils.decorators import apply_defaults
 from airflow.utils.operator_helpers import context_to_airflow_vars
 
 
+def _is_bash_script(input_string: str):
+    input_list = input_string.split(" ")
```
Comment (ashb) [inline airflow/operators/bash.py:34]: Maybe use shlex module?

File: airflow/operators/bash.py [hunk 2]
```diff
@@ -157,14 +163,37 @@ def pre_exec():
...
 
-            self.sub_process = Popen(  # pylint: disable=subprocess-popen-preexec-fn
-                ['bash', "-c", self.bash_command],
-                stdout=PIPE,
-                stderr=STDOUT,
-                cwd=tmp_dir,
-                env=env,
-                preexec_fn=pre_exec,
-            )
+            if isinstance(self.bash_command) == str and not _is_bash_script(self.bash_command):
```
Comment (ashb) [inline airflow/operators/bash.py:172]:
```suggestion
            if isinstance(self.bash_command, str) and _is_bash_script(self.bash_command):
```

File: airflow/operators/bash.py [hunk 3]
```diff
@@ -157,14 +163,37 @@ def pre_exec():
...
+                )
+            elif isinstance(self.bash_command) == list:
+                if (
+                    len(self.bash_command) >= 2
+                    and self.bash_command[0] == "bash"
+                    and self.bash_command[1] == "-c"
+                ):
+                    warnings.warn(
+                        "Warning: using \"bash -c\" for your bash command will give this command"
+                        "full access to the bash environment. Please consider not doing this."
```
Comment (ashb) [inline airflow/operators/bash.py]: I don't understand this warning.

File: airflow/operators/bash.py [hunk 4]
```diff
@@ -157,14 +163,37 @@ def pre_exec():
...
+                    cwd=tmp_dir,
+                    env=env,
+                    preexec_fn=pre_exec,
+                )
+            elif isinstance(self.bash_command) == list:
+                if (
+                    len(self.bash_command) >= 2
+                    and self.bash_command[0] == "bash"
+                    and self.bash_command[1] == "-c"
+                ):
```
Comment (ashb) [inline airflow/operators/bash.py]: Can't do multi line suggest on mobile app but

```bash_command[0:2] == ["bash", "-c"]```

**Review by ashb — CHANGES_REQUESTED** [review]
We also need to update docs and example dags

File: airflow/operators/bash.py [hunk 5]
```diff
@@ -19,16 +19,22 @@
...
 
 from airflow.exceptions import AirflowException
 from airflow.models import BaseOperator
 from airflow.utils.decorators import apply_defaults
 from airflow.utils.operator_helpers import context_to_airflow_vars
 
 
+def _is_bash_script(input_string: str):
+    input_list = input_string.split(" ")
+    return len(input_list) == 1 and len(input_string) > 2 and input_string[3:] == ".sh"
```
Comment (ashb) [inline airflow/operators/bash.py:35]: This check doesn't seem right it only also for `a.sh` not `longer.sh`

File: airflow/operators/bash.py [hunk 6]
```diff
@@ -95,10 +101,16 @@ class BashOperator(BaseOperator):
 
         bash_task = BashOperator(
             task_id="bash_task",
-            bash_command='echo "here is the message: \'$message\'"',
+            bash_command=['echo', "here is the message: \'$message\'"],
             env={'message': '{{ dag_run.conf["message"] if dag_run else "" }}'},
         )
 
+    We also highly recommend against using [\"bash\", \"-c\"... in your command as that command will negate
```
Comment (ashb) [inline airflow/operators/bash.py:108]: Don't need `\` here

File: airflow/operators/bash.py [hunk 7]
```diff
@@ -157,14 +163,31 @@ def pre_exec():
...
+                )
+                self.sub_process = Popen(  # pylint: disable=subprocess-popen-preexec-fn
+                    ['bash', "-c", self.bash_command],
+                    stdout=PIPE,
+                    stderr=STDOUT,
+                    cwd=tmp_dir,
+                    env=env,
+                    preexec_fn=pre_exec,
+                )
+            elif isinstance(self.bash_command) == list:
```
Comment (michalslowikowski00) [inline airflow/operators/bash.py:188]:
```suggestion
            elif isinstance(self.bash_command, list):
```

File: airflow/operators/bash.py (hunk 2 above)
Comment (kaxil) [reply airflow/operators/bash.py:172]: This PR needs some tests too @dimberman

File: airflow/operators/bash.py (hunk 4 above)
Comment (TheRook) [reply airflow/operators/bash.py]: This is good, they should get hit with a warning that is functionally equivalent and still leads to evaluation of bash entities.

File: airflow/operators/bash.py (hunk 6 above)
Comment (TheRook) [reply airflow/operators/bash.py:108]: Perhaps we can link to some docs here?  I am willing to write up a page on injection in general, and why parallelization is a good tool for the job.

File: airflow/operators/bash.py (hunk 4 above)
Comment (ashb) [reply airflow/operators/bash.py]: Yes, running bash is the point of the operator - it's called a BashOperetor, not ExecCommandOperator ;)

File: airflow/operators/bash.py (hunk 6 above)
Comment (ashb) [reply airflow/operators/bash.py:108]: Doc PRs are always appreciated!

**Comment by dstandish** [comment]
@dimberman this one may be of interest: https://github.com/apache/airflow/pull/13423

just stumbled on to the present pr... #13423 pulls the "run command" logic into a hook, and requires passing as list

**Comment by ashb** [comment]
@dstandish Yeah, we're going to close this, and create a separate operator -- leaving the Bash as _actually running bash_, and create a new one to called something like "RunCommandOperator" and make that the "default" in all the docs over BashOperator.

Ideas for a better/shorted name appreciated.

**Comment by dstandish** [comment]
> Ideas for a better/shorted name appreciated

well `CommandOperator` is shorter if not better :) 

there's also `SubprocessOperator`

or `PopenOperator`

**Comment by dstandish** [comment]
Lastly `ExecOperator`

**Comment by ashb** [comment]
> > Ideas for a better/shorted name appreciated
> 
> well `CommandOperator` is shorter if not better :)
> 
> there's also `SubprocessOperator`

> 
> or `PopenOperator`

Popen probably only makes sense if you are a python developer and know what Popen is.

Of those I think CommandOperator (or ExecOperator as you had before you edited it?).

**Comment by kaxil** [comment]
Personally `RunCommandOperator` feels a bit better

**Comment by ashb** [comment]
I wonder if we could have a "short" form version that fits in with the taskflow api:

```python
@dag
def my_workflow():

    @task
    def munge_output(line):
        return line.upper()

    date = run_command("date --iso=s")
    munge_output(date)
```

**Pull request closed without merging**